candle-metal-kernels/src/lib.rs

```diff
@@ -0,0 +1,1317 @@
+    encoder.dispatch_thread_groups(thread_group_count, thread_group_size);
+    encoder.end_encoding();
+    Ok(())
+}
+
+pub fn call_where_cond_strided(
+    device: &Device,
+    command_buffer: &CommandBufferRef,
+    kernels: &Kernels,
+    name: &'static str,
+    shape: &[usize],
+    cond: &Buffer,
+    (cond_stride, cond_offset): (&[usize], usize),
+    left: &Buffer,
+    (left_stride, left_offset): (&[usize], usize),
+    right: &Buffer,
+    (right_stride, right_offset): (&[usize], usize),
+    output: &mut Buffer,
+) -> Result<(), MetalKernelError> {
+    let func = kernels.load_function(device, Source::Ternary, name)?;
+    let pipeline_state_descriptor = ComputePipelineDescriptor::new();
+    pipeline_state_descriptor.set_compute_function(Some(&func));
+
+    let pipeline = device
+        .new_compute_pipeline_state_with_function(
+            pipeline_state_descriptor.compute_function().unwrap(),
+        )
+        .unwrap();
+
+    let encoder = command_buffer.new_compute_command_encoder();
+    encoder.set_compute_pipeline_state(&pipeline);
+
+    let size: usize = shape.iter().product();
+    let rank = shape.len();
+
+    set_params!(
+        encoder,
+        (
+            size,
+            rank,
+            shape,
+            cond_stride,
+            left_stride,
+            right_stride,
+            (cond, cond_offset),
+            (left, left_offset),
+            (right, right_offset),
+            output
+        )
+    );
+
+    let (thread_group_count, thread_group_size) = linear_split(&pipeline, size);
+
+    encoder.dispatch_thread_groups(thread_group_count, thread_group_size);
+    encoder.end_encoding();
+    Ok(())
+}
+
+pub fn call_index_select(
+    device: &Device,
+    command_buffer: &CommandBufferRef,
+    kernels: &Kernels,
+    name: &'static str,
+    shape: &[usize],
+    ids_size: usize,
+    dim: usize,
+    input: &Buffer,
+    ids: &Buffer,
+    output: &mut Buffer,
+) -> Result<(), MetalKernelError> {
+    let left_size: usize = shape[..dim].iter().product();
+    let right_size: usize = shape[dim + 1..].iter().product();
+    let src_dim_size = shape[dim];
+    let dst_el = ids_size * left_size * right_size;
+
+    let func = kernels.load_function(device, Source::Indexing, name)?;
+    let pipeline = device
+        .new_compute_pipeline_state_with_function(&func)
+        .unwrap();
+
+    let encoder = command_buffer.new_compute_command_encoder();
+
+    encoder.set_compute_pipeline_state(&pipeline);
+
+    set_params!(
+        encoder,
+        (
+            dst_el,
+            left_size,
+            src_dim_size,
+            right_size,
+            ids_size,
+            input,
+            ids,
+            output
+        )
+    );
+
+    let (thread_group_count, thread_group_size) = linear_split(&pipeline, dst_el);
+
+    encoder.dispatch_thread_groups(thread_group_count, thread_group_size);
+    encoder.end_encoding();
+    Ok(())
+}
+
+#[cfg(test)]
```

Comment:
Could you move the test in a separate file? For a large number of tests I find it much better to have them split out as it simplifies the review process.

Reply:
I prefer having tests close to their call site personally, makes imports easier and overall readability better.
It's also pretty much the standard in rust ecosystem: https://github.com/tokio-rs/tokio/blob/7a30504fd423aa782239ead92f3afa3474f8037c/tokio/src/util/memchr.rs#L29
https://github.com/serde-rs/serde/blob/2083f43a287cac8302009fda5bbe41518dd83209/test_suite/tests/test_ser.rs#L547
https://doc.rust-lang.org/book/ch11-01-writing-tests.html
https://github.com/BurntSushi/aho-corasick/blob/f227162f7c56b3f28e9dd69a655316af0107b29b/src/util/alphabet.rs#L300

For non `src/` tests, it also makes failure mode different (it fails fast by default).

I personnally don't have a strong preference tbh. I started moving but it start to feel less rusty, therefore I checked popular crates (above) and it seems to be definitely the norm to put the test in the same file as the logic they are testing.

Given this, would you reconsider?

Reply:
Slightly prefer the separate files as I find it much better when reviewing code to know if it's mostly test being affected vs the actual code so if you prefer it this way I'm fine with it (but would rather keep the rest of the tests separate as they currently are and have this only in `candle-metal-kernels`).

Reply:
Done.